python/ray/experimental/data/dataset.py

```diff
@@ -799,14 +799,78 @@ def iter_batches(self,
             prefetch_blocks: The number of blocks to prefetch ahead of the
                 current block during the scan.
             batch_size: Record batch size, or None to let the system pick.
-            batch_format: Specify "pandas" to select ``pandas.DataFrame`` as
-                the batch format, or "pyarrow" to select ``pyarrow.Table``.
+            batch_format: The format in which to return each batch.
+                Specify "pandas" to select ``pandas.DataFrame``, "pyarrow" to
+                select ``pyarrow.Table``, or "_blocks" to return the raw block.
+                Default is "pandas".
+            drop_last: Whether to drop the last batch if it's incomplete.
 
         Returns:
             A list of iterators over record batches.
         """
 
-        raise NotImplementedError  # P1
+        def window(iterable: Iterable, n: int):
+            iters = itertools.tee(iter(iterable), n)
+            for i in range(1, n):
+                for it in iters[i:]:
+                    next(it, None)
+            return zip(*iters)
+
+        def format_batch(batch: Block, format: str):
+            if batch_format == "pandas":
+                return batch.to_pandas()
+            elif batch_format == "pyarrow":
+                return batch._table
+            elif batch_format == "_blocks":
+                return batch
+            else:
+                raise ValueError(
+                    f"The given batch format: {batch_format} "
+                    f"is invalid. Supported batch type: {BatchType}")
+
+        batch_buffer = None
+        for block_window in window(self._blocks, prefetch_blocks + 1):
+            block_window = list(block_window)
+            # Prefetch blocks.
```

Comment:
This function is pretty long. Here's a cleanup suggestion: make a util class like this to make the high level logic more readable:

```
batcher = Batcher(batch_size=N)
for i, block in enumerate(blocks):
   ray.wait(blocks[i:i + num_prefetch])   # avoid itertools (principle of least power)
   batcher.add(block)
   while batcher.has_next_batch():
      yield batcher.next_batch()

if batcher.has_any() and not drop_last:
   yield batcher.leftover_batch()
```

Reply:
Could be like this
```
class Batcher:
    def __init__(self, batch_size):
        self.batch_size = batch_size
        self.buffer = []

    def add(self, batch):
        self.buffer.append(batch)

    def has_next(self):
        return sum(b.num_rows() for b in self.buffer) >= self.batch_size

    def next_batch(self):
        output = []
        leftover = []
        needed = self.batch_size

        for b in self.buffer:
            if needed <= 0:
                # Don't need
                leftover.append(b)
            elif b.num_rows() <= needed:
                # Need the entire batch
                leftover.append(b)
                output.append(b)
                needed -= b.num_rows()
            else:
                # Need the partial batch
                output.append(b.slice(0, needed))
                leftover.append(b.slice(needed, b.num_rows()))
                needed = 0

        self.buffer = leftover
        return concat(output)
```

Reply:
Agreed on isolating a lot of this logic to a `Batcher` iterator, and overall, I like consolidating the leftover batch buffer and the block iterator into the same buffer so the batch slicing and leftover handling is also consolidated.

Making a few notes mainly for my benefit for when I work on this refactor, and offering up an alternative `Batcher` that addresses a few of my concerns.

> `   ray.wait(blocks[i:i + num_prefetch])   # avoid itertools (principle of least power)`

The point of using itertools is to grab prefetching windows from the `BlockList` iterator without forcing creation of all blocks (e.g. with creating a list), which would break the laziness of `LazyBlockList` ([you left this very feedback](https://github.com/ray-project/ray/pull/16853#discussion_r663203529) on an earlier version of this PR). We can do the equivalent slicing on an iterator with `itertools.islice(i, i + num_prefetch)`, but that's just leaking the itertools implementation details that the `window()` utility is attempting to hide by providing a sliding window iterator.

> `        output = []`

I'm assuming that we'd still want to use the `DelegatingArrowBlockBuilder` as the output buffer.

```python
            elif b.num_rows() <= needed:
                # Need the entire batch
                leftover.append(b)
                output.append(b)
                needed -= b.num_rows()
```

Should only add to the output buffer here, not the leftover buffer.

Might also want an iterator helper:
```python
def __iter__(self):
    while self.has_batch():
        yield self.next_batch()
```


### Batcher concerns:

1. If a single block contains a lot of batches, we'll end up creating (a) more slices, and (b) larger slices. E.g., the following slices would be made for `block_size = 10` and `batch_size = 2`:
    - Current - 5 batch slices: slice(0, 2), slice(2, 4), slice(4, 6), slice(6, 8), slice(8, 10)
    - Proposed - 15 batch slices: slice(0, 2), slice(2, 10), slice(0, 2), slice(2, 8), slice(0, 2), slice(2, 6), slice(0, 2), slice(2, 4), slice(0, 2)
    - Not a big deal if these slices don't involve copies.
2. In general, the above `Batcher` seems geared towards batching across blocks, including being able to add `N` blocks before extracting batches. I believe that batching within blocks (`batch_size < block_size`) will be the more common usage, and I don't think that we'll need to add more than one block to the batcher before we'll want to try to extract and yield batches (YAGNI).
3. Making callers check if the batcher has more batches to return before calling `.next_batch()` is forcing a LBYL antipattern on the caller, when EAFP is more idiomatic Python (especially when the `StopIteration` raised by the generator is very naturally handled by Python for loops).

### Alternative batcher

Oriented towards `batch_size < block_size` rather than `batch_size > block_size`, and only needing to extract batches from one block at a time (with possible remainder from previous block(s)):

```python
batcher = Batcher(batch_size=batch_size)
for block_window in window(self._blocks, prefetch_blocks + 1):
    block_window = list(block_window)
    ray.wait(block_window, num_returns=1, fetch_local=False)
    block = ray.get(block_window[0])
    for batch in batcher.extract_batches(block):
        yield format(batch, batch_format)

if batcher.has_remainder_batch() and not drop_last:
   yield format(batcher.get_remainder_batch())


class Batcher:
    def __init__(self, batch_size):
        self.batch_size = batch_size
        self.leftover = DelegatingArrowBlockBuilder()

    def has_remainder_batch(self):
        return self.leftover.num_rows() > 0

    def extract_batches(self, block):
        batch_size = self.batch_size or block.num_rows()
        if self.has_remainder_batch():
            start = min(batch_size - self.leftover.num_rows(), block.num_rows())
            self.leftover.add_block(block.slice(0, start))
            if self.leftover.num_rows() == batch_size:
                yield self.leftover.build()
                self.leftover.clear()
        else:
            start = 0
        while start <= block.num_rows() - batch_size:
            yield block.slice(start, start + batch_size)
            start += batch_size
        if start < block.num_rows():
            self.leftover.add_block(block.slice(start, block.num_rows()))

    def get_remainder_batch(self):
        return self.leftover.build()
```

Reply:
For comparison, this is what your example would look like in full:
```python
batcher = Batcher(batch_size=batch_size)
for block_window in window(self._blocks, prefetch_blocks + 1):
    block_window = list(block_window)
    ray.wait(block_window, num_returns=1, fetch_local=False)
    block = ray.get(block_window[0])
    batcher.add(block)
    while batcher.has_batch():
        yield format(batcher.next_batch(), batch_format)

if batcher.has_remainder_batch() and not drop_last:
   yield format(batcher.get_remainder_batch())


class Batcher:
    def __init__(self, batch_size):
        self.batch_size = batch_size
        self.buffer = []

    def add(self, batch):
        self.buffer.append(batch)

    def has_batch(self):
        return sum(b.num_rows() for b in self.buffer) >= self.batch_size

    def has_any(self):
        return any(b.num_rows() > 0 for b in self.buffer)

    def next_batch(self):
        if self.batch_size is None:
            return concat(self.buffer)
        output = DelegatingArrowBlockBuilder()
        leftover = []
        needed = self.batch_size
        for b in self.buffer:
            if needed <= 0:
                # Don't need
                leftover.append(b)
            elif b.num_rows() <= needed:
                # Need the entire batch
                output.add_block(b)
                needed -= b.num_rows()
            else:
                # Need the partial batch
                output.add_block(b.slice(0, needed, copy=False))
                leftover.append(b.slice(needed, b.num_rows(), copy=False))
                needed = 0

        self.buffer = leftover
        return output.build()
```

Reply:
Can we do it the way I suggested? It's much more clear.

Also, I didn't realize slicing forced evaluation of the entire list. But yeah using `itertools.islice(i, i + num_prefetch)` directly sounds better than window(), since the reader can understand it directly without going to another function.

Reply:
>Can we do it the way I suggested? It's much more clear.

Sure, I'll make a note that the performance impact of the extra slicing and loops should be negligible since we're operating on zero-copy views of the blocks and that `block_size // batch_size` is not expected to be extremely large.

>But yeah using itertools.islice(i, i + num_prefetch) directly sounds better than window(), since the reader can understand it directly without going to another function.

I should have noted that `itertools.islice(iter(self._blocks), i, i + num_prefetch)` will recreate and re-consume the `iter(self._blocks)` iterator from the beginning and up to `i` on every iteration, which is very inefficient. I'd much rather go with a `sliding_window` utility with documented semantics and performance characteristics, that we can maybe reuse.

Reply:
Ok. Another option for the slicing is to implement slicing for LazyBlockList, so you can directly slice blocks without listing the entire thing.

Reply:
If we end up needing `BlockList.slice(i, j)` for other things, that sounds like a good idea! My guess is that the ideal `LazyBlockList.slice()` implementation would go something like:
```python
def slice(self, i: int, j: int):
    self._get_or_compute(j)
    return self._blocks[i : j]
```
, triggering the exponential rampup manually instead of through the iterator, the big pro here being that future slices wouldn't have to do a no-op iteration up through `i`.